cupyx/jit/_compile.py

```diff
@@ -546,18 +546,47 @@ def _transpile_expr_internal(expr, env):
     if isinstance(expr, ast.Subscript):
         value = _transpile_expr(expr.value, env)
         index = _transpile_expr(expr.slice, env)
-        if is_constants([value, index]):
-            return Constant(value[index])
+
+        if is_constants([value]):
+            if is_constants([index]):
+                return Constant(value.obj[index.obj])
+            raise TypeError(
+                f'{type(value.obj)} is not subscriptable with non-constants.')
+
         value = _to_cuda_object(value, env)
-        index = _to_cuda_object(index, env)
-
-        if not isinstance(value.ctype, _types.Array):
-            raise ValueError(f'{value.code} must be Array type.')
-        if value.ctype.ndim != 1:
-            raise NotImplementedError('Not implemented for ndim > 1.')
-        return CudaObject(
-            f'{value.code}[{index.code}]',
-            _types.Scalar(value.ctype.dtype))
+
+        if isinstance(value.ctype, _types.Tuple):
+            raise NotImplementedError
+
+        if isinstance(value.ctype, _types.Array):
+            index = _to_cuda_object(index, env)
+            ndim = value.ctype.ndim
+            if isinstance(index.ctype, _types.Scalar):
+                index_dtype = index.ctype.dtype
+                if ndim != 1:
+                    raise TypeError(
+                        'Scalar indexing is supported only for 1-dim array.')
+                if index_dtype.kind not in 'ui':
+                    raise TypeError('Array indices must be integers.')
+                return CudaObject(
+                    f'{value.code}[{index.code}]',
+                    _types.Scalar(value.ctype.dtype))
+            if isinstance(index.ctype, _types.Tuple):
+                if ndim != len(index.ctype.types):
+                    raise IndexError(f'The size of index must be {ndim}')
+                for t in index.ctype.types:
+                    if not isinstance(t, _types.Scalar):
+                        raise TypeError('Array indices must be scalar.')
+                    if t.dtype.kind not in 'iu':
+                        raise TypeError('Array indices must be integer.')
```

Comment:
I wonder if an empty tuple would hard error before arriving here.

Reply:
Empty tuples will be caught here:
https://github.com/cupy/cupy/pull/4939/files#diff-92c76e1dfae3dd9e376fac22b852d0deb6cea6d0affa13bda3ac7d2ee33ce17dR575-R576
, but indexing of 0-dim array is not tested. I will add the test in this PR.

Reply:
I noticed 0-dim array does not have `_indexing` member function. Fixed in cbf02f3!